test/preview/components/generators/openai/test_chatgpt_generator.py

```diff
@@ -0,0 +1,322 @@
+    return convert_to_openai_object(deepcopy(base_dict))
+
+
+def mock_openai_stream_response(messages: str, model: str = "gpt-3.5-turbo-0301", **kwargs) -> openai.ChatCompletion:
+    response = f"response for these messages --> {' - '.join(msg['role']+': '+msg['content'] for msg in messages)}"
+    base_dict = {
+        "id": "chatcmpl-7NaPEA6sgX7LnNPyKPbRlsyqLbr5V",
+        "object": "chat.completion",
+        "created": 1685855844,
+        "model": model,
+    }
+    base_dict["choices"] = [{"delta": {"role": "assistant"}, "finish_reason": None, "index": "0"}]
+    yield convert_to_openai_object(base_dict)
+    for token in response.split():
+        base_dict["choices"][0]["delta"] = {"content": token + " "}
+        yield convert_to_openai_object(base_dict)
+    base_dict["choices"] = [{"delta": {"content": ""}, "finish_reason": "stop", "index": "0"}]
+    yield convert_to_openai_object(base_dict)
+
+
+class TestChatGPTGenerator:
+    @pytest.mark.unit
+    def test_init_default(self, caplog):
+        component = ChatGPTGenerator()
+        assert component.system_prompt is None
+        assert component.api_key is None
+        assert component.model_name == "gpt-3.5-turbo"
+        assert component.streaming_callback is None
+        assert component.api_base_url == "https://api.openai.com/v1"
+        assert component.model_parameters is None
+        assert (
+            caplog.records[0].message == "OpenAI API key is missing. You need to provide an API key to Pipeline.run()."
+        )
+
+    @pytest.mark.unit
+    def test_init_with_parameters(self, caplog):
+        callback = lambda x: x
+        component = ChatGPTGenerator(
+            api_key="test-api-key",
+            model_name="gpt-4",
+            system_prompt="test-system-prompt",
+            model_parameters={"max_tokens": 10, "some-test-param": "test-params"},
+            streaming_callback=callback,
+            api_base_url="test-base-url",
+        )
+        assert component.system_prompt == "test-system-prompt"
+        assert component.api_key == "test-api-key"
+        assert component.model_name == "gpt-4"
+        assert component.streaming_callback == callback
+        assert component.api_base_url == "test-base-url"
+        assert component.model_parameters == {"max_tokens": 10, "some-test-param": "test-params"}
+        assert not caplog.records
+
+    @pytest.mark.unit
+    def test_to_dict_default(self):
+        component = ChatGPTGenerator()
+        data = component.to_dict()
+        assert data == {
+            "type": "ChatGPTGenerator",
+            "init_parameters": {
+                "api_key": None,
+                "model_name": "gpt-3.5-turbo",
+                "system_prompt": None,
+                "model_parameters": None,
+                "streaming_callback": None,
+                "api_base_url": "https://api.openai.com/v1",
+            },
+        }
+
+    @pytest.mark.unit
+    def test_to_dict_with_parameters(self):
```

Comment:
Let's also test a `to_dict` with a lambda callback.

Reply:
That would not be very meaningful, right? The lambda can't really be deserialized. Maybe I misunderstood.

Reply:
Added anyway, at least for documentation of the behavior. Later we can discuss how to properly serialize such functions and if it makes sense to do so.